[quant][graphmode] Support quantizing shared ClassType with different qconfigs

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32379 [WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes
* #32543 [WIP][quant][graphmode] ConvBn Folding works with bias=None
* #32380 [quant][graphmode][refactor] Better API for fold_convbn
* #32374 [refactor] Adding FoldConvBatchNorm2dHelper
* #32303 [quant][graphmode] Call _jit_pass_dedup_module_ueses in quantize_script
* #32302 [quant] Re-enable fold_convbn in quantize_script
* #32206 [quant] Re-enable test_nested that has different qconfig for shared ClassType
* **#32205 [quant][graphmode] Support quantizing shared ClassType with different qconfigs**

Summary:
Previously we don't support quantizing modules with shared ClassTypes but with different qconfigs. For example, conv1 and conv2 both sharing the same Conv2d ClassType, but we want to quantize conv1 and not conv2, then we need to produce two different quantized convs.
To achieve this, we customized `clone` to take the qconfig of the instance into consideration when deciding whether to create a new type or not. That is, the custom clone function will produce different types for the cloned conv1 and conv2 since they have different qconfigs(one is None and the other is not), therefore we can quantize them differently.

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19508031](https://our.internmc.facebook.com/intern/diff/D19508031)

jerryzh168 committed:
[quant][graphmode] Support quantizing shared ClassType with different qconfigs

Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 196047f0:


* **8/8** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=196047f03912bd1e49316b440a0c698c406bafa9).

### :detective: 8 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4248303) pytorch_xla_linux_xenial_py3_6_clang7_build (1/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86223002) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86223002))
<details>
<summary>
<code>Jan 15 02:41:17 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 15 02:41:17 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 15 02:41:17 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 15 02:41:17 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 15 02:41:17 1 out of 1 hunk FAILED 
Jan 15 02:41:17 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 15 02:41:17 Traceback (most recent call last): 
Jan 15 02:41:17   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 15 02:41:17     patch_repo(args) 
Jan 15 02:41:17   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 15 02:41:17     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 15 02:41:17 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 15 02:41:17 + cleanup 
Jan 15 02:41:17 + retcode=1 
Jan 15 02:41:17 + set +x 
Jan 15 02:41:17 =================== sccache compilation log =================== 
Jan 15 02:41:17 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 15 02:41:17 Compile requests                 1 
Jan 15 02:41:17 Compile requests executed        0 
Jan 15 02:41:17 Cache hits                       0 
Jan 15 02:41:17 Cache misses                     0 
Jan 15 02:41:17 Cache timeouts                   0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4248321) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (2/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86223531) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86223531))
<details>
<summary>
<code>Jan 15 02:43:21 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:890:33: error: call to function 'operator==' that is neither visible in the template definition nor found by argument-dependent lookup</code>
</summary>

```
Jan 15 02:43:10 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Jan 15 02:43:15 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/fuse_linear.cpp.o 
Jan 15 02:43:21 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:1: 
Jan 15 02:43:21 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.h:8: 
Jan 15 02:43:21 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/ir.h:3: 
Jan 15 02:43:21 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/attributes.h:2: 
Jan 15 02:43:21 In file included from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:3: 
Jan 15 02:43:21 In file included from /var/lib/jenkins/workspace/c10/core/Allocator.h:4: 
Jan 15 02:43:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/memory:79: 
Jan 15 02:43:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:55: 
Jan 15 02:43:21 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:890:33: error: call to function 'operator==' that is neither visible in the template definition nor found by argument-dependent lookup 
Jan 15 02:43:21         return bool(std::get<__i>(__t) == std::get<__i>(__u)) 
Jan 15 02:43:21                                        ^ 
Jan 15 02:43:21 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:923:25: note: in instantiation of member function 'std::__tuple_compare<std::tuple<torch::jit::script::Module, torch::jit::script::Module>, std::tuple<torch::jit::script::Module, torch::jit::script::Module>, 0, 2>::__eq' requested here 
Jan 15 02:43:21       return __compare::__eq(__t, __u); 
Jan 15 02:43:21                         ^ 
Jan 15 02:43:21 /var/lib/jenkins/workspace/c10/util/Optional.h:633:68: note: in instantiation of function template specialization 'std::operator==<torch::jit::script::Module, torch::jit::script::Module, torch::jit::script::Module, torch::jit::script::Module>' requested here 
Jan 15 02:43:21   return bool(x) != bool(y) ? false : bool(x) == false ? true : *x == *y; 
Jan 15 02:43:21                                                                    ^ 
Jan 15 02:43:21 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/stl_function.h:357:20: note: in instantiation of function template specialization 'c10::operator==<std::tuple<torch::jit::script::Module, torch::jit::script::Module> >' requested here 
Jan 15 02:43:21       { return __x == __y; } 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4248316) pytorch_linux_xenial_py3_clang5_asan_build (3/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86225529) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86225529))
<details>
<summary>
<code>Jan 15 02:44:03 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:890:33: error: call to function 'operator==' that is neither visible in the template definition nor found by argument-dependent lookup</code>
</summary>

```
Jan 15 02:44:03 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/register_prim_ops.cpp.o 
Jan 15 02:44:03 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/register_prim_ops_c10.cpp.o 
Jan 15 02:44:03 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:1: 
Jan 15 02:44:03 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.h:8: 
Jan 15 02:44:03 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/ir.h:3: 
Jan 15 02:44:03 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/attributes.h:2: 
Jan 15 02:44:03 In file included from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:3: 
Jan 15 02:44:03 In file included from /var/lib/jenkins/workspace/c10/core/Allocator.h:4: 
Jan 15 02:44:03 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/memory:79: 
Jan 15 02:44:03 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:55: 
Jan 15 02:44:03 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:890:33: error: call to function 'operator==' that is neither visible in the template definition nor found by argument-dependent lookup 
Jan 15 02:44:03         return bool(std::get<__i>(__t) == std::get<__i>(__u)) 
Jan 15 02:44:03                                        ^ 
Jan 15 02:44:03 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:923:25: note: in instantiation of member function 'std::__tuple_compare<std::tuple<torch::jit::script::Module, torch::jit::script::Module>, std::tuple<torch::jit::script::Module, torch::jit::script::Module>, 0, 2>::__eq' requested here 
Jan 15 02:44:03       return __compare::__eq(__t, __u); 
Jan 15 02:44:03                         ^ 
Jan 15 02:44:03 /var/lib/jenkins/workspace/c10/util/Optional.h:633:68: note: in instantiation of function template specialization 'std::operator==<torch::jit::script::Module, torch::jit::script::Module, torch::jit::script::Module, torch::jit::script::Module>' requested here 
Jan 15 02:44:03   return bool(x) != bool(y) ? false : bool(x) == false ? true : *x == *y; 
Jan 15 02:44:03                                                                    ^ 
Jan 15 02:44:03 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/stl_function.h:357:20: note: in instantiation of function template specialization 'c10::operator==<std::tuple<torch::jit::script::Module, torch::jit::script::Module> >' requested here 
Jan 15 02:44:03       { return __x == __y; } 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4248313) pytorch_macos_10_13_py3_build (4/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86227781) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86227781))
<details>
<summary>
<code>Jan 14 18:45:41 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o</code>
</summary>

```
Jan 14 18:45:30 [2491/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/subgraph_rewrite.cpp.o 
Jan 14 18:45:30 [2492/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/python_print.cpp.o 
Jan 14 18:45:31 [2493/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/subgraph_utils.cpp.o 
Jan 14 18:45:31 [2494/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/check_alias_annotation.cpp.o 
Jan 14 18:45:31 [2495/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/memory_dag.cpp.o 
Jan 14 18:45:39 [2496/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/fuse_linear.cpp.o 
Jan 14 18:45:40 [2497/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/print_handler.cpp.o 
Jan 14 18:45:40 [2498/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/fuser/interface.cpp.o 
Jan 14 18:45:41 [2499/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/register_prim_ops.cpp.o 
Jan 14 18:45:41 [2500/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Jan 14 18:45:41 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o  
READ -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -c ../torch/csrc/jit/passes/quantization.cpp 
Jan 14 18:45:41 In file included from ../torch/csrc/jit/passes/quantization.cpp:1: 
Jan 14 18:45:41 In file included from ../torch/csrc/jit/passes/quantization.h:8: 
Jan 14 18:45:41 In file included from ../torch/csrc/jit/ir.h:3: 
Jan 14 18:45:41 In file included from ../torch/csrc/jit/attributes.h:2: 
Jan 14 18:45:41 In file included from ../aten/src/ATen/ATen.h:3: 
Jan 14 18:45:41 In file included from ../c10/core/Allocator.h:4: 
Jan 14 18:45:41 In file included from /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:653: 
Jan 14 18:45:41 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/tuple:1126:77: error: invalid operands to binary expression ('const typename tuple_element<0UL, tuple<Module, Module> >::type' (aka 'const __type_pack_element<0UL, torch::jit::script::Module, torch::jit::script::Module>') and 'const typename tuple_element<0UL, tuple<Module, Module> >::type') 
Jan 14 18:45:41         return __tuple_equal<_Ip - 1>()(__x, __y) && _VSTD::get<_Ip-1>(__x) == _VSTD::get<_Ip-1>(__y); 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4248309) pytorch_macos_10_13_cuda10_0_cudnn7_py3_build (5/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86229172) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86229172))
<details>
<summary>
<code>Jan 14 18:48:28 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o</code>
</summary>

```
Jan 14 18:48:25 [2517/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/edit_distance.cpp.o 
Jan 14 18:48:25 [2518/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/logging.cpp.o 
Jan 14 18:48:26 [2519/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/module.cpp.o 
Jan 14 18:48:26 [2520/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/object.cpp.o 
Jan 14 18:48:26 [2521/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/jit_exception.cpp.o 
Jan 14 18:48:27 [2522/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/string_to_type.cpp.o 
Jan 14 18:48:27 [2523/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/source_range_serialization.cpp.o 
Jan 14 18:48:27 [2524/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/tracer.cpp.o 
Jan 14 18:48:28 [2525/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/hooks_for_testing.cpp.o 
Jan 14 18:48:28 [2526/2988] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Jan 14 18:48:28 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o  
READ -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -c ../torch/csrc/jit/passes/quantization.cpp 
Jan 14 18:48:28 In file included from ../torch/csrc/jit/passes/quantization.cpp:1: 
Jan 14 18:48:28 In file included from ../torch/csrc/jit/passes/quantization.h:8: 
Jan 14 18:48:28 In file included from ../torch/csrc/jit/ir.h:3: 
Jan 14 18:48:28 In file included from ../torch/csrc/jit/attributes.h:2: 
Jan 14 18:48:28 In file included from ../aten/src/ATen/ATen.h:3: 
Jan 14 18:48:28 In file included from ../c10/core/Allocator.h:4: 
Jan 14 18:48:28 In file included from /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:653: 
Jan 14 18:48:28 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/tuple:1126:77: error: invalid operands to binary expression ('const typename tuple_element<0UL, tuple<Module, Module> >::type' (aka 'const __type_pack_element<0UL, torch::jit::script::Module, torch::jit::script::Module>') and 'const typename tuple_element<0UL, tuple<Module, Module> >::type') 
Jan 14 18:48:28         return __tuple_equal<_Ip - 1>()(__x, __y) && _VSTD::get<_Ip-1>(__x) == _VSTD::get<_Ip-1>(__y); 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4248306) pytorch_linux_xenial_py3_clang5_android_ndk_r19c_mobile_code_analysis (6/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86235770) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86235770))
<details>
<summary>
<code>Jan 15 02:58:55 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o</code>
</summary>

```
Jan 15 02:58:40         if (*type->dim() >= num_reduced_dim && num_reduced_dim > 0) { 
Jan 15 02:58:40             ~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~ 
Jan 15 02:58:40 1 warning generated. 
Jan 15 02:58:41 [566/613] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/subgraph_rewrite.cpp.o 
Jan 15 02:58:43 [567/613] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/memory_dag.cpp.o 
Jan 15 02:58:50 [568/613] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/subgraph_utils.cpp.o 
Jan 15 02:58:52 [569/613] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/check_alias_annotation.cpp.o 
Jan 15 02:58:53 [570/613] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/print_handler.cpp.o 
Jan 15 02:58:53 [571/613] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/python_print.cpp.o 
Jan 15 02:58:55 [572/613] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Jan 15 02:58:55 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o  
rameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -c /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp 
Jan 15 02:58:55 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:1: 
Jan 15 02:58:55 In file included from ../../torch/csrc/jit/passes/quantization.h:8: 
Jan 15 02:58:55 In file included from ../../torch/csrc/jit/ir.h:3: 
Jan 15 02:58:55 In file included from ../../torch/csrc/jit/attributes.h:2: 
Jan 15 02:58:55 In file included from ../../aten/src/ATen/ATen.h:3: 
Jan 15 02:58:55 In file included from ../../c10/core/Allocator.h:4: 
Jan 15 02:58:55 In file included from /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/memory:79: 
Jan 15 02:58:55 In file included from /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:55: 
Jan 15 02:58:55 /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:890:33: error: call to function 'operator==' that is neither visible in the template definition nor found by argument-dependent lookup 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4248297) pytorch_linux_xenial_py3_clang5_android_ndk_r19c_x86_32_build (7/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86237545) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86237545))
<details>
<summary>
<code>Jan 15 03:02:25 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o</code>
</summary>

```
Jan 15 03:02:03 [533/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/requires_grad_analysis.cpp.o 
Jan 15 03:02:03 [534/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/specialize_autogradzero.cpp.o 
Jan 15 03:02:10 [535/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/subgraph_rewrite.cpp.o 
Jan 15 03:02:10 [536/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/shape_analysis.cpp.o 
Jan 15 03:02:16 [537/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/subgraph_utils.cpp.o 
Jan 15 03:02:18 [538/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/memory_dag.cpp.o 
Jan 15 03:02:23 [539/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/check_alias_annotation.cpp.o 
Jan 15 03:02:24 [540/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/python_print.cpp.o 
Jan 15 03:02:24 [541/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/print_handler.cpp.o 
Jan 15 03:02:25 [542/583] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Jan 15 03:02:25 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o  
rameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -c /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp 
Jan 15 03:02:25 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:1: 
Jan 15 03:02:25 In file included from ../torch/csrc/jit/passes/quantization.h:8: 
Jan 15 03:02:25 In file included from ../torch/csrc/jit/ir.h:3: 
Jan 15 03:02:25 In file included from ../torch/csrc/jit/attributes.h:2: 
Jan 15 03:02:25 In file included from ../aten/src/ATen/ATen.h:3: 
Jan 15 03:02:25 In file included from ../c10/core/Allocator.h:4: 
Jan 15 03:02:25 In file included from /opt/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/memory:663: 
Jan 15 03:02:25 /opt/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/tuple:1122:77: error: invalid operands to binary expression ('const typename tuple_element<0U, tuple<Module, Module> >::type' (aka 'const __type_pack_element<0U, torch::jit::script::Module, torch::jit::script::Module>') and 'const typename tuple_element<0U, tuple<Module, Module> >::type') 
Jan 15 03:02:25         return __tuple_equal<_Ip - 1>()(__x, __y) && _VSTD::get<_Ip-1>(__x) == _VSTD::get<_Ip-1>(__y); 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4248233) binary_macos_libtorch_2_7_cpu_build (8/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86259390) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86259390))
<details>
<summary>
<code>Jan 15 03:31:44 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o</code>
</summary>

```
Jan 15 03:31:32 ../torch/csrc/jit/passes/shape_analysis.cpp:1244:26: warning: comparison of integers of different signs: 'unsigned long' and 'int64_t' (aka 'long long') [-Wsign-compare] 
Jan 15 03:31:32         if (*type->dim() >= num_reduced_dim && num_reduced_dim > 0) { 
Jan 15 03:31:32             ~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~ 
Jan 15 03:31:32 1 warning generated. 
Jan 15 03:31:33 [2505/3005] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/memory_dag.cpp.o 
Jan 15 03:31:36 [2506/3005] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/subgraph_utils.cpp.o 
Jan 15 03:31:40 [2507/3005] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/check_alias_annotation.cpp.o 
Jan 15 03:31:41 [2508/3005] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/python_print.cpp.o 
Jan 15 03:31:41 [2509/3005] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/print_handler.cpp.o 
Jan 15 03:31:44 [2510/3005] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Jan 15 03:31:44 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o  
READ -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -c ../torch/csrc/jit/passes/quantization.cpp 
Jan 15 03:31:44 In file included from ../torch/csrc/jit/passes/quantization.cpp:1: 
Jan 15 03:31:44 In file included from ../torch/csrc/jit/passes/quantization.h:8: 
Jan 15 03:31:44 In file included from ../torch/csrc/jit/ir.h:3: 
Jan 15 03:31:44 In file included from ../torch/csrc/jit/attributes.h:2: 
Jan 15 03:31:44 In file included from ../aten/src/ATen/ATen.h:3: 
Jan 15 03:31:44 In file included from ../c10/core/Allocator.h:4: 
Jan 15 03:31:44 In file included from /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:653: 
Jan 15 03:31:44 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/tuple:1126:77: error: invalid operands to binary expression ('const typename tuple_element<0UL, tuple<Module, Module> >::type' (aka 'const __type_pack_element<0UL, torch::jit::script::Module, torch::jit::script::Module>') and 'const typename tuple_element<0UL, tuple<Module, Module> >::type') 
Jan 15 03:31:44         return __tuple_equal<_Ip - 1>()(__x, __y) && _VSTD::get<_Ip-1>(__x) == _VSTD::get<_Ip-1>(__y); 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
Overall looks good, please don't forget to clang-format.

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -179,6 +179,173 @@ Value* getModuleAccessPath(Value* instance, std::vector<std::string>& path) {
+    }
+
+    // only clone the methods if the ClassType is not cloned before
+    if (!type_already_cloned) {
```
I tried before, but we need to clone the methods after we add attributes, which is done after initialization..

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -179,6 +179,173 @@ Value* getModuleAccessPath(Value* instance, std::vector<std::string>& path) {
+      const ModuleQConfigMap& module_qconfig_map,
+      const std::unordered_map<TypePtr, QConfigTypePtrMap>& type_remap) {
+    // type remapping - when we copy method implementations from one module
+    // singleton to another, we need to update the types of the self arguments
```
this is copied from previous code :)

jerryzh168 replied:
I'll remove this comment

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1198,7 +1365,10 @@ TORCH_API script::Module InsertObservers(
+  ModuleCloneHelper mh;
+  script::Module module = inplace ? input_module : mh.clone(input_module, map_before_clone);
   ModuleQConfigMap module_qconfig_map;
   fillQConfigMap(module, qconfig_dict, module_qconfig_map);
```
yes, reason is that the types are changed after clone

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
to be filled

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support quantizing shared ClassType with different qconfigs"


Summary:
Previously we don't support quantizing modules with shared ClassTypes but with different qconfigs. For example, conv1 and conv2 both sharing the same Conv2d ClassType, but we want to quantize conv1 and not conv2, then we need to produce two different quantized convs.
To achieve this, we customized `clone` to take the qconfig of the instance into consideration when deciding whether to create a new type or not. That is, the custom clone function will produce different types for the cloned conv1 and conv2 since they have different qconfigs(one is None and the other is not), therefore we can quantize them differently.

Test Plan:
python test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19508031](https://our.internmc.facebook.com/intern/diff/D19508031)

[ghstack-poisoned]
